src/server/protocol.d.ts

```diff
@@ -0,0 +1,820 @@
+/** 
+  * Declaration module describing the TypeScript Server protocol 
+  */
+declare module ts.server.protocol {
+    /** 
+      * A TypeScript Server message 
+      */
+    export interface Message {
+        /** 
+          * Sequence number of the message 
+          */
+        seq: number;
+
+        /**
+          * One of "request", "response", or "event" 
+          */
+        type: string;
+    }
+    
+    /** 
+      * Client-initiated request message 
+      */
+    export interface Request extends Message {
+        /**
+          * The command to execute 
+          */
+        command: string;
+
+        /** 
+          * Object containing arguments for the command 
+          */
+        arguments?: any;
+    }
+
+    /** 
+      * Server-initiated event message 
+      */
+    export interface Event extends Message {
+        /** 
+          * Name of event 
+          */
+        event: string;
+
+        /** 
+          * Event-specific information 
+          */
+        body?: any;
+    }
+
+    /**
+      * Response by server to client request message.
+      */
+    export interface Response extends Message {
+        /**
+          * Sequence number of the request message.
+          */
+        request_seq: number;
+
+        /** 
+          * Outcome of the request. 
+          */
+        success: boolean;
+
+        /** 
+          * The command requested.
+          */
+        command: string;
+
+        /** 
+          * Contains error message if success == false. 
+          */
+        message?: string;

```

Comment:
Why not do it like how we do the shim layer.  Always just have a body, but hte body can have "error", "canceled" or "result"?